### ISSUE TITLE
undefined method `simple_name' for nil:NilClass (NoMethodError)

### ISSUE THREAD

**Issue by backus** [troessner/reek#715]
The following two examples produce an error:

```ruby
CONST = new
```

```ruby
class Foo
  INSTANCE = new
end
```

this is similar to #708 but not the same.

Just to double check, I cloned reek and ran it against these examples:

```sh
$ git clone https://github.com/troessner/reek.git
Cloning into 'reek'...
remote: Counting objects: 14309, done.
remote: Compressing objects: 100% (107/107), done.
remote: Total 14309 (delta 23), reused 0 (delta 0), pack-reused 14202
Receiving objects: 100% (14309/14309), 3.75 MiB | 3.55 MiB/s, done.
Resolving deltas: 100% (7372/7372), done.
Checking connectivity... done.
$ cd reek
$ git log | head -n1
Tue Sep 22 09:32:25 2015 +0200 3c2d85f (HEAD -> master, origin/master, origin/HEAD) Merge pull request #700 from troessner/fix-smell-details-handling-in-reek-of-matcher  [Matijs van Zuijlen]
$ cd ..
$ cat example1.rb
CONST = new
$ cat example2.rb
class Foo
  INSTANCE = new
end
$ reek/bin/reek example1.rb
/private/tmp/reek/lib/reek/ast/sexp_extensions.rb:435:in `defines_module?': undefined method `simple_name' for nil:NilClass (NoMethodError)
	from /private/tmp/reek/lib/reek/tree_walker.rb:60:in `process_casgn'
	from /private/tmp/reek/lib/reek/tree_walker.rb:44:in `process'
	from /private/tmp/reek/lib/reek/tree_walker.rb:38:in `result'
	from /private/tmp/reek/lib/reek/tree_walker.rb:27:in `walk'
	from /private/tmp/reek/lib/reek/examiner.rb:87:in `run'
	from /private/tmp/reek/lib/reek/examiner.rb:42:in `initialize'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:13:in `new'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:13:in `block in execute'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:12:in `each'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:12:in `execute'
	from /private/tmp/reek/lib/reek/cli/application.rb:31:in `execute'
	from reek/bin/reek:12:in `<main>'
$ reek/bin/reek example2.rb
/private/tmp/reek/lib/reek/ast/sexp_extensions.rb:435:in `defines_module?': undefined method `simple_name' for nil:NilClass (NoMethodError)
	from /private/tmp/reek/lib/reek/tree_walker.rb:60:in `process_casgn'
	from /private/tmp/reek/lib/reek/tree_walker.rb:44:in `process'
	from /private/tmp/reek/lib/reek/tree_walker.rb:83:in `block in process_default'
	from /private/tmp/reek/lib/reek/tree_walker.rb:82:in `each'
	from /private/tmp/reek/lib/reek/tree_walker.rb:82:in `process_default'
	from /private/tmp/reek/lib/reek/tree_walker.rb:53:in `block in process_module'
	from /private/tmp/reek/lib/reek/tree_walker.rb:200:in `block in inside_new_context'
	from /private/tmp/reek/lib/reek/tree_walker.rb:213:in `push'
	from /private/tmp/reek/lib/reek/tree_walker.rb:199:in `inside_new_context'
	from /private/tmp/reek/lib/reek/tree_walker.rb:52:in `process_module'
	from /private/tmp/reek/lib/reek/tree_walker.rb:44:in `process'
	from /private/tmp/reek/lib/reek/tree_walker.rb:38:in `result'
	from /private/tmp/reek/lib/reek/tree_walker.rb:27:in `walk'
	from /private/tmp/reek/lib/reek/examiner.rb:87:in `run'
	from /private/tmp/reek/lib/reek/examiner.rb:42:in `initialize'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:13:in `new'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:13:in `block in execute'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:12:in `each'
	from /private/tmp/reek/lib/reek/cli/reek_command.rb:12:in `execute'
	from /private/tmp/reek/lib/reek/cli/application.rb:31:in `execute'
	from reek/bin/reek:12:in `<main>'
```

Note that both the constant and `new` seem to be necessary:

```sh
$ cat example3.rb
class Foo
  instance = new
end
$ cat example4.rb
class Foo
  @@instance = new
end
$ cat example5.rb
CONST = foo
$ reek/bin/reek example3.rb
$ reek/bin/reek example4.rb
example4.rb -- 1 warning:
  [2]:Foo declares the class variable @@instance (ClassVariable)
$ reek/bin/reek example5.rb
$
```

The issue popped up at least 17 days ago (I found out about it through mbj/devtools#84).  To see this error in context, see [lib/morpher.rb#L102](https://github.com/mbj/morpher/blob/master/lib/morpher.rb#L102)

---

Bisecting seems to point to b4eea25276090cdc00ef3c4dc3aed7c95cffcd10:

```sh
$ git bisect start master 5875ead
Already on 'master'
Your branch is up-to-date with 'origin/master'.
Bisecting: 86 revisions left to test after this (roughly 7 steps)
[ad623574a0410610ea4c8501695ea4da1fee2b17] Merge pull request #661 from troessner/increase-rubocops-line-length
$ git bisect run ./bin/reek ../example1.rb 2> /dev/null
running ./bin/reek ../example1.rb
Bisecting: 42 revisions left to test after this (roughly 6 steps)
[3a3d64f327fb67ee9c63758d738f2e5877a1de83] Merge pull request #635 from troessner/improve-readme-tools-integrations-and-authors-reloaded
running ./bin/reek ../example1.rb
Bisecting: 21 revisions left to test after this (roughly 5 steps)
[248adcd5e5a06e355b2f77bfefaf46a80c1177f8] Merge pull request #607 from troessner/update-change-log
running ./bin/reek ../example1.rb
Bisecting: 10 revisions left to test after this (roughly 4 steps)
[b4eea25276090cdc00ef3c4dc3aed7c95cffcd10] Handle modules defined by constant assignment
running ./bin/reek ../example1.rb
Bisecting: 4 revisions left to test after this (roughly 3 steps)
[c7fcc23e6955c83461857be922fa07e1ef63bb98] Merge pull request #596 from troessner/switch_to_pathname
running ./bin/reek ../example1.rb
Bisecting: 2 revisions left to test after this (roughly 1 step)
[28e53468b9c5d213a9fdfc9b89111247d9e6d9bf] Merge pull request #598 from troessner/drop_method_to_proc
running ./bin/reek ../example1.rb
Bisecting: 0 revisions left to test after this (roughly 1 step)
[9956353f5488fda2b80c665498cd94b88bcc5579] Merge pull request #590 from troessner/module-function-modifier
running ./bin/reek ../example1.rb
b4eea25276090cdc00ef3c4dc3aed7c95cffcd10 is the first bad commit
commit b4eea25276090cdc00ef3c4dc3aed7c95cffcd10
Author: Matijs van Zuijlen <matijs@matijs.net>
Date:   Sun Jul 12 09:02:12 2015 +0200

    Handle modules defined by constant assignment

    Introduces a new ModuleContext for classes defined by constant
    assignment, like so:

        Foo = Class.new do
          def bar(a, b)
          end
        end

    This means smells on methods inside the class definition will have the
    correct context reported.

    Also makes IrresponsibleModule detect when such classes have no comment.
    Other smells still have to be adjusted to consider :casgn contexts.

:040000 040000 afdf750a40e5385e81234d0e55cf143ea4eb54c2 96d50e49bdd89b511a50295f9a4a4fbd25926fdf M	lib
:040000 040000 9a282449e6054a5003ff729f5ef17bb1ab72ec90 365c29381d1089f88eabeb6600312305f680b4bb M	spec
bisect run success
```

**Comment by chastell** [comment]
Thanks for the bug report! I’ll have a fix for this in a sec.

**Comment by chastell** [comment]
This is now fixed in the freshly-released reek 3.4.1 – thanks for the bug report!
